src/Tools/dotnet-dsproxy/DiagnosticServerProxyCommands.cs

```diff
@@ -0,0 +1,98 @@
+            using CancellationTokenSource linkedCancelToken = CancellationTokenSource.CreateLinkedTokenSource(token, cancelProxyTask.Token);
+
+            var proxyTask = DiagnosticServerProxyFactory.runIpcClientTcpServerProxy(linkedCancelToken.Token, ipcClient, tcpServer, autoShutdown, debug);
+
+            while (!linkedCancelToken.IsCancellationRequested)
+            {
+                await Task.WhenAny(proxyTask, Task.Delay(250)).ConfigureAwait(false);
+                if (proxyTask.IsCompleted)
+                    break;
+
+                if (Console.KeyAvailable)
+                {
+                    ConsoleKey cmd = Console.ReadKey(true).Key;
+                    if (cmd == ConsoleKey.Q)
+                    {
+                        cancelProxyTask.Cancel();
+                        break;
+                    }
+                }
+            }
+
+            return proxyTask.Result;
+        }
+
+        public async Task<int> RunIpcServerTcpServerProxy(CancellationToken token, string ipcServer, string tcpServer, bool autoShutdown, bool debug)
+        {
+            checkLoopbackOnly(tcpServer);
+
+            using CancellationTokenSource cancelProxyTask = new CancellationTokenSource();
+            using CancellationTokenSource linkedCancelToken = CancellationTokenSource.CreateLinkedTokenSource(token, cancelProxyTask.Token);
+
+            var proxyTask = DiagnosticServerProxyFactory.runIpcServerTcpServerProxy(linkedCancelToken.Token, ipcServer, tcpServer, autoShutdown, debug);
+
+            while (!linkedCancelToken.IsCancellationRequested)
+            {
+                await Task.WhenAny(proxyTask, Task.Delay(250)).ConfigureAwait(false);
+                if (proxyTask.IsCompleted)
+                    break;
+
+                if (Console.KeyAvailable)
+                {
+                    ConsoleKey cmd = Console.ReadKey(true).Key;
+                    if (cmd == ConsoleKey.Q)
+                    {
+                        cancelProxyTask.Cancel();
+                        break;
+                    }
+                }
+            }
+
+            return proxyTask.Result;
+        }
+
+        static void checkLoopbackOnly(string tcpServer)
+        {
+            if (!DiagnosticServerProxyFactory.isLoopbackOnly(tcpServer))
+            {
+                StringBuilder message = new StringBuilder();
+
+                message.Append("WARNING: Binding tcp server endpoint to anything except loopback interface ");
+                message.Append("(localhost, 127.0.0.1 or [::1]) is NOT recommended. Any connections towards ");
+                message.Append("tcp server endpoint will be unauthenticated and unencrypted. This component ");
+                message.Append("is intented for development use and should only be run in development and ");
+                message.Append("testing environments.");
+                message.AppendLine();
```

Comment:
@shirhatti do you have any thoughts on the phrasing of this warning? This tool is _only_ intended for people to use in a very specific development time scenario. Is this strongly worded enough?

Reply:
I'm okay with this message.

As an aside though, I'd love for us to do an end-to-end review with the runtime changes and dsproxy.

@blowdart Would probably appreciate the  heads-up that we'll have something for him 😅

Reply:
Don't worry 😉  It's already in the works by @lateralusX

Reply:
@shirhatti, @josalem Any thoughts about the tool name. Right now I call it dotnet-dsproxy but maybe dotnet-dsrouter or just dotnet-router is a better one since tool mainly routes diagnostic server ipc<->tcp traffic between mobile devices and diagnostic tooling.

Reply:
Switched name to dotnet-dsrouter.